tests/ui/consts/invalid-union.64bit.stderr

```diff
@@ -6,7 +6,7 @@ LL | fn main() {
    |
    = note: The rules on what exactly is undefined behavior aren't clear, so this check might be overzealous. Please open an issue on the rustc repository if you believe it should not be considered undefined behavior.
    = note: the raw bytes of the constant (size: 8, align: 8) {
-               ╾───────alloc7────────╼                         │ ╾──────╼
+               ╾───────alloc8────────╼                         │ ╾──────╼
```

Comment:
Hm, AllocIds are going up a bit... is the round-trip via valtrees leading do allocations that we somehow managed to avoid before?

Reply:
yea, valtree roundtrip often creates new allocations

Reply:
valtree_to_const_value has a shortcut for some very trivial types, but previously we'd also use that for e.g. ADTs with scalar ABI. If this shows up as a perf regression I guess I can try to make it kick in for more types.

Reply:
Hm, strangely adding an optimization for Scalar structs changes absolutely nothing about the AllocIds. I guess we are round-tripping larger types or reference, though I cannot see where these tests would have a ty::Const that's not a primitive (non-reference) type...

Reply:
The `valtree_to_const_val` codepath in `mir::ConstantKind::eval` isn't even entered in these tests where the AllocId increases...

Reply:
I found the source of this... if I change `eval_mir_constant` in the interpreter to this, the AllocIds revert to what they were:
```rust
    pub fn eval_mir_constant(
        &self,
        val: &mir::ConstantKind<'tcx>,
        span: Option<Span>,
        layout: Option<TyAndLayout<'tcx>>,
    ) -> InterpResult<'tcx, OpTy<'tcx, M::Provenance>> {
        match *val {
            mir::ConstantKind::Ty(ct) => {
                let ty = ct.ty();
                let valtree = self.ctfe_query(span, |tcx| ct.eval(*tcx, self.param_env, span))?;
                let const_val = self.tcx.valtree_to_const_val((ty, valtree));
                self.const_val_to_op(const_val, ty, layout)
            }
            mir::ConstantKind::Val(val, ty) => self.const_val_to_op(val, ty, layout),
            mir::ConstantKind::Unevaluated(uv, _) => {
                let instance = self.resolve(uv.def, uv.args)?;
                Ok(self.eval_global(GlobalId { instance, promoted: uv.promoted }, span)?.into())
            }
        }
    }
```
So it's completely unrelated to new valtree roundtrips. Instead it's somehow about evaluating a constant via `eval_to_allocation_raw` (inside `eval_global`) vs via `const_eval_resolve` (inside `eval`). And really strangely, it is the  `const_eval_resolve`  that leads to more AllocId, despite the fact that it is `eval_to_allocation_raw` which forces an allocation!

`const_eval_resolve` calls `eval_to_const_value_raw` and that in turn calls `eval_to_allocation_raw`, so... they should be sharing the same underlying query. But somehow they are not?

Reply:
Somehow it's about how the result of `eval_to_allocation_raw` is processed.

If I pass it directly to `raw_const_to_mplace`, we get the old behavior (lower AllocIds). If instead I call `turn_into_const_value` and then `const_val_to_op`, we get higher AllocIds.

Aha! `const_val_to_op` for an `ConstValue::ByRef` needs to reserve an AllocId to get a pointer to that const. The same const data might already have an AllocId, but we don't have a cache for that direction (we do intern the underlying `ConstAllocation` data).

So yeah it's a bit silly. We are not evaluating any constants twice (phew), but we do assign multiple AllocId to the same data -- first in `eval_to_allocation_raw`, which returns an `AllocId`, and then again in `const_val_to_op` when the `ConstAllocation` needs to be turned back into an `AllocId`.

So... what could we do? `eval_to_allocation_raw` always internally has an AllocId, so if we want to keep the two representations consistent then `ConstValue::ByRef` should also have an AllocId instead of a `ConstAllocation`. Basically it would be an `interpret::Pointer` at that point, `AllocId` + offset. Or alternatively we cache the mapping from `ConstAllocation` to the `AllocId` that contains them. Or we do nothing. But honestly just using a `Pointer` in `ConstValue::ByRef` seems like the easiest and most consistent option.

Reply:
Yea, but I have a vague feeling that we tried that before and perf was unhappy. But this may not be true anymore today (since we have valtrees).

Reply:
Digging through codegen it looks like just having an AllocId in `ByRef` won't deduplicate anything, `const_data_from_alloc` still gets called each time we encounter a `ByRef` constant. We'd have to add a cache somewhere for the mapping from `AllocId` to codegen backend values. That might or might not be actually helpful for performance.